### PR TITLE
Upgrade all Python dependencies

### PR DESCRIPTION
**Pull request by nwalters512**
Done with the following using [`pur`](https://pypi.org/project/pur/):

```sh
pur -r graders/python/requirements.txt
pur -r graders/c/requirements.txt
pur -r workspaces/jupyterlab-python/requirements.txt
pur -r images/plbase/python-requirements.txt
```

I then manually reverted the bumps for `pandas`, `pygraphviz`, `rpy2`, and `tzlocal`. I also reverted `lxml` and `lxml-stubs` in `images/plbase`.

### CODE REVIEW

**Comment by SethPoulsen** [comment]
What's with all the unrelated Python changes? Did `black` introduce not-backward compatible changes? or can we avoid those changes through a config change?

I'm happy to say the packages updated won't affect my python code, I'm not sure if I can say anything about other parts of the codebase though, which makes me nervous about approving.

File: images/plbase/python-requirements.txt [hunk 1]
```diff
@@ -1,47 +1,47 @@
 ansi2html==1.9.1
-black==23.12.1
+black==24.1.1
```
Comment (eliotwrobson) [inline images/plbase/python-requirements.txt:2]: Saw the Python changes and figured this got bumped.

File: images/plbase/python-requirements.txt [hunk 2]
```diff
@@ -1,47 +1,47 @@
...
+pycryptodome==3.20.0
 Pygments==2.17.2
 pygments-ansi-color==0.3.0
 pygraphviz==1.6
 pyquaternion==0.9.9
 pytest-cov==4.1.0
 pytest-lazy-fixture==0.6.3
 pytest-repeat==0.9.3
-pytest==7.4.4
+pytest==8.0.0
```
Comment (eliotwrobson) [inline images/plbase/python-requirements.txt:31]: This is the only major version bump to be a little wary about, but instructor code doesn't use it so probably not an issue.

File: images/plbase/python-requirements.txt [hunk 3]
```diff
@@ -1,47 +1,47 @@
...
+black==24.1.1
 chevron==0.14.0
 coloraide==2.16
-Faker==22.0.0
+Faker==22.6.0
 flake8-black==0.3.6
 flake8-isort==6.1.1
 Flake8-pyproject==1.2.3
-flake8==6.1.0
+flake8==7.0.0
```
Comment (eliotwrobson) [inline images/plbase/python-requirements.txt:9]: Like the other major version bump, shouldn't affect any course code.

**Review by eliotwrobson — APPROVED** [review]
LGTM

File: images/plbase/python-requirements.txt (hunk 2 above)
Comment (jonatanschroeder) [reply images/plbase/python-requirements.txt:31]: > instructor code doesn't use it

That's a bit of a dangerous assumption. I use pytest in a few of my questions (though in the autograder, not in server.py). That said, I agree that pytest use in plbase instructor code is expected to be uncommon.

Comment (jonatanschroeder) [reply images/plbase/python-requirements.txt:31]: Also, I see the possibility that custom elements could be implemented with pytest checks in some scenarios for dev checks, though I don't foresee a case where the inconvenience of a possible (and likely uncommon) breaking change in dev work would be a reason not to update here.

**Review by jonatanschroeder — APPROVED** [review]

**Comment by nwalters512** [comment]
@eliotwrobson it looks like the `pytest` changes might have broken us, here's an excerpt from our CI logs:

```
==================================== ERRORS ====================================
_ ERROR collecting apps/prairielearn/elements/pl-big-o-input/big_o_utils_test.py _
ImportError while importing test module '/home/runner/work/PrairieLearn/PrairieLearn/apps/prairielearn/elements/pl-big-o-input/big_o_utils_test.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
apps/prairielearn/elements/pl-big-o-input/big_o_utils_test.py:1: in <module>
    import big_o_utils as bou
apps/prairielearn/elements/pl-big-o-input/big_o_utils.py:3: in <module>
    import python_helper_sympy as phs
E   ModuleNotFoundError: No module named 'python_helper_sympy'
____ ERROR collecting apps/prairielearn/elements/pl-code/code_utils_test.py ____
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_hooks.py:501: in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_manager.py:1[19](https://github.com/PrairieLearn/PrairieLearn/actions/runs/7717814440/job/21037725784?pr=9311#step:15:20): in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/python.py:277: in pytest_pycollect_makeitem
    return list(collector._genfunctions(name, obj))
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/_pytest/python.py:506: in _genfunctions
    self.ihook.pytest_generate_tests.call_extra(methods, dict(metafunc=metafunc))
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_hooks.py:562: in call_extra
    return self._hookexec(self.name, hookimpls, kwargs, firstresult)
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pluggy/_manager.py:119: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pytest_lazyfixture.py:74: in pytest_generate_tests
    normalize_metafunc_calls(metafunc, 'funcargs')
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pytest_lazyfixture.py:81: in normalize_metafunc_calls
    calls = normalize_call(callspec, metafunc, valtype, used_keys)
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/pytest_lazyfixture.py:105: in normalize_call
    valtype_keys = set(getattr(callspec, valtype).keys()) - used_keys
E   AttributeError: 'CallSpec2' object has no attribute 'funcargs'
_ ERROR collecting apps/prairielearn/elements/pl-number-input/pl_number_input_test.py _
ImportError while importing test module '/home/runner/work/PrairieLearn/PrairieLearn/apps/prairielearn/elements/pl-number-input/pl_number_input_test.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
apps/prairielearn/elements/pl-number-input/pl_number_input_test.py:5: in <module>
    number_input = importlib.import_module("pl-number-input")
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/importlib/__init__.py:1[26](https://github.com/PrairieLearn/PrairieLearn/actions/runs/7717814440/job/21037725784?pr=9311#step:15:27): in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
apps/prairielearn/elements/pl-number-input/pl-number-input.py:8: in <module>
    import prairielearn as pl
E   ModuleNotFoundError: No module named 'prairielearn'
_ ERROR collecting apps/prairielearn/elements/pl-units-input/units_utils_test.py _
ImportError while importing test module '/home/runner/work/PrairieLearn/PrairieLearn/apps/prairielearn/elements/pl-units-input/units_utils_test.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
apps/prairielearn/elements/pl-units-input/units_utils_test.py:5: in <module>
    import unit_utils as uu
apps/prairielearn/elements/pl-units-input/unit_utils.py:5: in <module>
    import prairielearn as pl
E   ModuleNotFoundError: No module named 'prairielearn'
```

Do you have the bandwidth to look into this? If not, I'm inclined to revert the `pytest` upgrade for now.

**Comment by eliotwrobson** [comment]
@nwalters512 I actually noticed similar errors on a pytest upgrade on a different project I have, and couldn't figure out what was going on then. Go ahead and revert and I'll report back when I figure out what's up.

**Pull request merged**